src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java

```diff
@@ -48,24 +59,52 @@ public PreviewsView(EventBus eventBus) {
     }
 
     @Subscribe
+    /**
+     * This function is called when a preview button is pushed/triggered
+     */
     public synchronized void onSocketPreviewChanged(SocketPreviewChangedEvent event) {
-        Platform.runLater(() -> {
-            final OutputSocket<?> socket = event.getSocket();
-
-            if (socket.isPreviewed()) {
-                // If the socket was just set as previewed, add it to the list of previewed sockets and add a new view for it.
-                if (!this.previewedSockets.contains(socket)) {
-                    this.previewedSockets.add(socket);
-                    this.previewBox.getChildren().add(SocketPreviewViewFactory.createPreviewView(this.eventBus, socket));
-                }
-            } else {
-                // If the socket was just set as not previewed, remove both it and the corresponding control
-                int index = this.previewedSockets.indexOf(socket);
-                if (index != -1) {
-                    this.previewedSockets.remove(index);
-                    this.eventBus.unregister(this.previewBox.getChildren().remove(index));
+
+        final OutputSocket<?> socket = event.getSocket(); //The socket whose preview has changed
+
+        if (socket.isPreviewed()) {// If the socket was just set as previewed, add it to the list of previewed sockets and add a new view for it.
+
+            if (!this.previewedSockets.contains(socket)) {//If the socket is not already previewed...
+
+                if (socket.getStep().isPresent()) { //If this is a socket associated with a pipeline step (IE NOT a source)....
+
+                    Step socketStep = socket.getStep().get();//The pipeline step associated with the socket whose preview has changed
+                    final StepView stepView = this.pipeline.findStepView(socketStep);//The gui object that displays the socketStep
+                    int indexOfStep = this.pipeline.getSteps().indexOf(stepView); //The index of the step that has the socket in the pipeline
+
+                    int numbOfSourcePreviews = 0;//This we will use to count how many *source* previews (not *step* previews) are currently displayed
+                    while ((this.previewedSockets.size() > numbOfSourcePreviews) //While there are still previews to examine
+                            && ((this.previewedSockets.get(numbOfSourcePreviews).getSocketHint().getIdentifier().contains("Image")) //If this is a source (currently, sources can only have two types of output sockets, "Image" and "Frame Rate")...
+                            || (this.previewedSockets.get(numbOfSourcePreviews).getSocketHint().getIdentifier().contains("Frame Rate"))))
```

Comment:
I don't like that this is a hard coded conditional off of a constant variable declared in another file.